utils/datastructures/set.go

```diff
@@ -0,0 +1,54 @@
+package datastructures
+
+import "sync"
+
+type SyncSet[T comparable] struct {
+	dict map[T]bool
+	mu   *sync.Mutex
+}
+
+func NewSyncSet[T comparable](initial []T) SyncSet[T] {
+	res := SyncSet[T]{
+		dict: map[T]bool{},
+		mu:   &sync.Mutex{},
+	}
+	for _, s := range initial {
+		res.dict[s] = true
+	}
+	return res
+}
+
+func (s *SyncSet[T]) Add(val T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.dict[val] = true
+}
+
+func (s *SyncSet[T]) Remove(val T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	delete(s.dict, val)
+}
+
+func (s *SyncSet[T]) RemoveAll(vals []T) {
+	for _, val := range vals {
+		s.Remove(val)
+	}
+}
+
+func (s *SyncSet[T]) Contains(val T) bool {
+	_, ok := s.dict[val]
+	return ok
+}
+
+func (s *SyncSet[T]) ToSlice() []T {
```

Comment:
made it sorted. This was like this because the only place that calls it doesn't care about ordering, but agreed that we should sort it to prevent future use cases from breaking